File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        hs = set()

        n = len(s)
        l,r = 0,0 
        maxi = 0
        while r<n:
            if s[r] not in hs:
                hs.add(s[r])
                r += 1
            else:
                hs.remove(s[l])
                l += 1
            maxi = max(maxi, r-l)
        
        return maxi

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<>();
        int j = 0, maxi = 0;
        for(int i=0;i< s.length();i++){
            while(j<i && set.contains(s.charAt(i))){
                set.remove(s.charAt(j));
                j++;
            }
            set.add(s.charAt(i));
            maxi = Math.max(i-j+1,maxi);
        }
        return maxi;
    }
}